### DISCUSSION TITLE
How to use this sceneForm for AugmentedFace?

### DISCUSSION THREAD

**Discussion by zjafarian** [sceneview/sceneform-android#72] [Q&A]
HI
I develop makeup by augmented face using SceneView, but  AugmentedFaceNode is not for rendering, can you help me?

**Comment by ThomasGorisse** [comment]
Did you try the 1.1.5 AugmentedFace sample ?

**Reply by zjafarian** [reply]
No, I didn't see or find the 1.1.5 AugmentedFace sample, could you send a link? thank you

**Comment by ThomasGorisse** [comment]
The 1.1.5 sample is here : [v1.15.0/sceneform-android-sdk/samples/augmentedfaces/](https://github.com/google-ar/sceneform-android-sdk/tree/v1.15.0/samples/augmentedfaces)

@grassydragon @Oleur @issacclee @fvito @issacclee Anyone who want to have fun moving the 1.15.0 AugmentedFaceNode and AngmentedFaces sample to master?

It shouldn't be very difficult but I actually don't know why the `AugmentedFaceNode` class was removed on the 1.16.0

**Reply by ThomasGorisse** [reply]
And I have to finish that before: [Filament logo 3D Model](https://github.com/google/filament/discussions/4010) to include latest materials extensions usage in a sample (They are handled by Filament and have a very nice looking)

If anyone wants a beginner course about using Blender for AR usage, I can drive you.

**Reply by zjafarian** [reply]
oh, I saw this sample, in this sample rendering by OpenGL but I wanted to use sceneform(AugmentedFaceNode) that AugmentedFaceNode class was removed.

I would be happy if there is a beginner course about AR and everything about Ar  :)

**Reply by grassydragon** [reply]
Hi!
I would recommend to first read the ARCore documentation. It has a good [introduction to AR](https://developers.google.com/ar/discover) and [design guidelines](https://developers.google.com/ar/design). Also, you can take a look at the [ARCore Elements](https://play.google.com/store/apps/details?id=com.google.ar.unity.ddelements) application that shows these guidelines in action. For tutorials you can check out this [website](https://www.raywenderlich.com/). The tutorials are paid there but have a good quality. I would still recommend you to read about OpenGL to understand how everything works internally. For interactive tutorials in the browser you can look at the [WebGL Fundamentals](https://webglfundamentals.org/) since WebGL and OpenGL have many things in common.

**Reply by zjafarian** [reply]
Hi
Thanks for your guide :) 👍

**Reply by fvito** [reply]
@ThomasGorisse Hi, I completely missed this discussion, but I can take a look at creating a sample for the face tracking if no one else is working on it,, seems fairly straight forward

**Reply by ThomasGorisse** [reply]
It will be great. Thanks a lot.

**Reply by fvito** [reply]
I think I've figured out why `AugmentedFaceNode` is not in public sceneform release. I suspect it's because it uses proprietary code or assets (like the `face_mesh` and face_`occluder_mesh`) or they were not able to convert it from sfb. 
They also seemed to have removed the `SkeletonNode` which is an important part of the `AugmentedFaceNode` (in order to translate the mesh being projected on the face).

So far what I've managed to figure out is, augmented faces example is showing 2 parts.
1.) Projecting mesh on top of the face
2.) Projecting face texture onto the face

At first, I tried to port just the face texture projection, but I'm now a bit skeptical that is possible. That example uses `sceneform_face_mesh` and `sceneform_face_mesh_occluder` which are not availabile in the public sceneform repo. I've tried to take the canonical_face_mesh.fbx and convert it to glb, but it seems to be broken, as the submesh count is returned as 0 (which then breaks the material loading) - something that has been reported to their repository.

I'm still attempting to maybe atleast port the mesh projection, but its not as easy as I thought :slightly_smiling_face: 
From what I understand, I need to take the face regions and apply their position to the bones of the projected model.

**Reply by ThomasGorisse** [reply]
Can you send me the fbx you found?
I'll try to have a look at what is inside.
If you have any .obj, I'm also interested.

For the bone position there is the ‘FilamentInstance.setBonesMatrix()‘ from what I know. You can maybe have a look at this.

**Reply by fvito** [reply]
At first I tried using mesh from [here](https://github.com/google-ar/arcore-android-sdk/tree/master/assets) as that one looks more  like a face. but have now found the "meshes" that I was missing in obj format [here](https://github.com/google-ar/sceneform-android-sdk/tree/master/sceneformux/ux/sampledata) but when converting them to glb with blender, I still get submesh error. Although I'm not sure if they are correct, since they seem to just be a simple plane. Also, does the material also get converted or do I need to convert it with matc?

Another obstacle I encountered is, getting the actuall bone/joints from the models. The sample `fox_face` and the guide on AR Core, expects the assets to be a single mesh with 4 joints. In order to position them correctly, we would need to get them from the file. I tried looking at how the glb is read/loaded but was only able to find a buffer. I assume this is done in the gltfio library. `Renderable.getRenderableInstance.getFillamentAssets.Entities` does contain them, but I was not able find a way to get the actuall entity, EntityManager only supports creating and destroying them. The only references to any skeleton/bones seems to be related to SFB file that was being used before.

**Reply by fvito** [reply]
I have some good news.
I was able to make the facemesh material/occluder to work. This means we can atleast project different textures into the face. For the other part of the sample (projecting mesh) I'm still looking at how to support bones.
I've found a sorta way to get the bones via transform manager, will play around with that a bit. Worse case, atleast face texture now works

Also, sorry for spamming here

**Comment by ThomasGorisse** [comment]
Hi @fvito,

My apologies for not responding sooner I was working on the 1.19.1

So, to summarize, if I well understood, what you only missing is a way to retrieve the bones matrix from the glb file (FilamentInstance) in order to map it with the ARCore face matrix?

Could you make the Pull Request in order to help you with that?

Don't worry by the fact that it is still not finalize. The PR will be Squashed and merged at the end.
That means that it will create a branch on which you can commit all and as much as you want on it and at the merge time, every commit on that branch will become a single commit on the master branch.

Thanks

**Reply by fvito** [reply]
Hi, no worries :slightly_smiling_face: 

I've just opened the PR with the ported sample. I the end I did not use the bones, as even with just basic translation/rotation it behaved the same as their sample so I'm not sure we need bone translation.
I'm facing another issue though, where the camera stream stopped rendering when using front facing camera. I've written more about it in the PR description. If anyone has any idea what change might have caused it, I would appriciate any sort of help.
